File: MergeSort/MergeSort.java
public class MergeSort {
    public static void merge(int[] array, int[] left, int[] right, int leftSize, int rightSize) {
        int i = 0, j = 0, k = 0;
        while(i < leftSize && j < rightSize) {
            if(left[i] < right[j]) array[k++] = left[i++];
            else array[k++] = right[j++];
        }
        while(i < leftSize) array[k++] = left[i++];
        while(j < rightSize) array[k++] = right[j++];
    }

    public static void mergeSort(int[] array) {
        if(array.length < 2) return;
        int mid = array.length / 2;
        int[] left = new int[mid], right = new int[array.length - mid];
        for(int i = 0; i < mid; i++) left[i] = array[i];
        for(int i = mid; i < array.length; i++) right[i - mid] = array[i];
        mergeSort(left);
        mergeSort(right);
        merge(array, left, right, mid, array.length - mid);
    }

    public static void printArray(String arrayName, int[] array) {
        System.out.print(arrayName + " = { ");
        for(int i = 0; i < array.length; i++) {
            System.out.print(array[i]);
            if(i != array.length - 1) System.out.print(", ");
        }
        System.out.println(" }");
    }

    public static void main(String[] args) {
        int[] numberArray = {45, 86, 34, 50, 24, 57, 41, 21};
        printArray("UNSORTED ARRAY", numberArray);
        // UNSORTED ARRAY = { 45, 86, 34, 50, 24, 57, 41, 21 }
        mergeSort(numberArray);
        printArray("SORTED ARRAY", numberArray);
        // SORTED ARRAY = { 21, 24, 34, 41, 45, 50, 57, 86 }
    }
}

File: MergeSort/mergeSort.cpp
#include <iostream>

template<typename T>
void merge(T array[], T left[], int leftSize, T right[], int rightSize) {
    int i = 0, j = 0, k = 0;
    while(i < leftSize && j < rightSize) {
        if(left[i] < right[j]) array[k++] = left[i++];
        else array[k++] = right[j++];
    }
    while(i < leftSize) array[k++] = left[i++];
    while(j < rightSize) array[k++] = right[j++];
}

template<typename T>
void mergeSort(T array[], int arraySize) {
    if(arraySize < 2) return;
    int mid = arraySize / 2;
    T left[mid], right[arraySize - mid];
    for(int i = 0; i < mid; i++) left[i] = array[i];
    for(int i = mid; i < arraySize; i++) right[i - mid] = array[i];
    mergeSort(left, mid);
    mergeSort(right, arraySize - mid);
    merge(array, left, mid, right, arraySize - mid);
}

template<typename T>
void printArray(char arrayName[], T array[], int arraySize) {
    std::cout << arrayName << " = { ";
    for(int i = 0; i < arraySize; i++) {
        std::cout << array[i];
        if(i != arraySize - 1) std::cout << ", ";
    }
    std::cout << " }" << std::endl;
}

int main() {
    int numberArray[] = {87, 74, 89, 76, 50, 71, 21, 90};
    int arraySize = sizeof(numberArray) / sizeof(numberArray[0]);
    printArray("UNSORTED ARRAY", numberArray, arraySize);
    // UNSORTED ARRAY = { 87, 74, 89, 76, 50, 71, 21, 90 }
    mergeSort(numberArray, arraySize);
    printArray("SORTED ARRAY", numberArray, arraySize);
    // SORTED ARRAY = { 21, 50, 71, 74, 76, 87, 89, 90 }
    return 0;
}

